services/motion/builtin/state/state.go

```diff
@@ -52,16 +67,529 @@ type PlannerExecutor interface {
+	return cs.executionIDHistory[0]
+}
+
+// execution represents the state of a motion planning execution.
+// it only ever exists in state.StartExecution function & the go routine created.
+type execution[R any] struct {
+	id                      motion.ExecutionID
+	state                   *State
+	waitGroup               *sync.WaitGroup
+	cancelCtx               context.Context
+	cancelFunc              context.CancelFunc
+	logger                  logging.Logger
+	componentName           resource.Name
+	req                     R
+	planExecutorConstructor PlanExecutorConstructor[R]
+}
+
+type planWithExecutor struct {
+	plan         motion.Plan
+	planExecutor PlannerExecutor
+	waypoints    Waypoints
+	motionplan   motionplan.Plan
+}
+
+// NewPlan creates a new motion.Plan from an execution & returns an error if one was not able to be created.
+func (e *execution[R]) newPlanWithExecutor(seedPlan motionplan.Plan, replanCount int) (planWithExecutor, error) {
+	pe, err := e.planExecutorConstructor(e.cancelCtx, e.req, seedPlan, replanCount)
+	if err != nil {
+		return planWithExecutor{}, err
+	}
+	resp, err := pe.Plan()
+	if err != nil {
+		return planWithExecutor{}, err
+	}
+
+	plan := motion.Plan{
+		ID:            uuid.New(),
+		ExecutionID:   e.id,
+		ComponentName: e.componentName,
+		Steps:         resp.PosesByComponent,
+	}
+	return planWithExecutor{plan: plan, planExecutor: pe, waypoints: resp.Waypoints, motionplan: resp.Motionplan}, nil
+}
+
+// Start starts an execution with a given plan.
+func (e *execution[R]) start() error {
+	var replanCount int
+	originalPlanWithExecutor, err := e.newPlanWithExecutor(nil, replanCount)
+	if err != nil {
+		return err
+	}
+	e.notifyStateNewExecution(e.toStateExecution(), originalPlanWithExecutor.plan, time.Now())
+	// We need to add to both the state & execution waitgroups
+	// B/c both the state & the stateExecution need to know if this
+	// goroutine have termianted.
+	// state.Stop() needs to wait for ALL execution goroutines to terminate before
+	// returning in order to not leak.
+	// Similarly stateExecution.stop(), which is called by state.StopExecutionByResource
+	// needs to wait for its 1 execution go routine to termiante before returning.
+	// As a result, both waitgroups need to be written to.
+	e.state.waitGroup.Add(1)
+	e.waitGroup.Add(1)
+
+	utils.PanicCapturingGo(func() {
+		defer e.state.waitGroup.Done()
+		defer e.waitGroup.Done()
+
+		lastPWE := originalPlanWithExecutor
+		// Exit conditions of this loop:
+		// 1. The execution's context was cancelled, which happens if the state's Stop() was called or
+		// StopExecutionByResource was called for this resource
+		// 2. the execution succeeded
+		// 3. the execution failed
+		// 4. replanning failed
+		for {
+			resChan := make(chan struct {
+				resp ExecuteResponse
+				err  error
+			}, 1)
+			utils.PanicCapturingGo(func() {
+				replan, err := lastPWE.planExecutor.Execute(lastPWE.waypoints)
```

Comment:
Is the motivation behind not making the execute function handle the spinning up of the goroutine that it becomes easier to test?